Allow separation and lazy loading of vendor folder in s3

tomhatzer opened the pull request:
Dear Bref team,

this is my first big contribution to this project and I hope this one will become a small part of this great project.

**WHY**
I'm currently trying out a few bigger opensource eCommerce/CMS/PIM scripts based on Symfony in my spare time and was curious if I could get them deployed on Lambda. The Lambda .zip archive went up to 80 megabytes and more due to the amount of packages in the vendor directory and immediately reached the limits of AWS Lambda.

Together with @mnapoli, who was a great helper, I've created this modification that allows bref users to store the vendor folder archive in s3 and then lazy load it into the Lambda temporary directory. There are also modifications in place for composer to load the correct source files.

Based on the input from @mnapoli we did not extend the layers with additional binaries and instead used PHP code to do the s3 downloading and ext-zip to do the unpacking of the vendor archive.

I'm looking forward to your comments and ideas on this.

Thank you very much and have a great day!

Best wishes
Tom

flemzord commented on docs/environment/php.md:
````diff
@@ -177,3 +177,14 @@ provider:
+
+```yaml
+custom:
+    separateVendor: true
````
Add level, like this ? 

```yaml
custom:
    bref:
        separateVendor: true
```

mnapoli replied:
I agree, this would be better to "namespace" bref features as they may be others in the future.

tomhatzer replied:
I've added `bref` to the namespace like you recommended 😃

flemzord commented:
Retrieving vendors from S3 + the unzip will explode the ColdStart :( 
Wouldn't it be simpler to have the vendors in an EFS for example ? Which can be mounted directly in a Lambda.

tomhatzer commented:
@flemzord Thanks for your comments!

Yes, that's true. Keeping the layers clean from additional binaries was the goal in this PR. Doing the downloading and extracting stuff with php sadly doesn't bring any performance benefit at all. I haven't tried doing it using the binaries, some people who tried said it wouldn't take longer than a few ms. With PHP it takes about 3-4 seconds.

I'm not sure if it would make things simpler as the setup would have to deal with stuff that's not yet built by serverless itself (we are using the serverless bucket for vendor file storage) and it would also push costs higher and add ~35$/month for the NAT gateway just to get Lambda to EFS connectivity + additional transfer costs from and to the NAT gateway.

mnapoli commented:
> Retrieving vendors from S3 + the unzip will explode the ColdStart :(

@flemzord Could you justify this please?

For the record AWS Lambda fetches your application from zip from S3 on cold start. So we're doing the same thing here. Theoretically speaking, this shouldn't be slower. Of course, reality may be different.

Note:

- I'm not saying this method should be used by everyone, only for those really limited by the Lambda max size, and looking to avoid messing with Docker (people who could not use Bref/Lambda before)
- I'm not saying this method is faster/slower/as fast for cold starts, I'm just saying "let's not make assumptions"

> Wouldn't it be simpler to have the vendors in an EFS for example ? Which can be mounted directly in a Lambda.

I benchmarked that: expect 30s+ cold starts. Things may have changed since then, you are welcome to benchmark that again.

mnapoli reviewed (commented):
😅  went through it, that's huge work, thank you!

I added some questions and comments

(edit: just noticed I forgot to click "submit review" this morning, sorry!)

tomhatzer commented on index.js:
```diff
@@ -34,19 +39,210 @@ class ServerlessPlugin {
+        this.newVendorZipName = vendorZipHash + '.zip';
+
+        let excludes = this.serverless.service.package.exclude;
+        if(excludes.indexOf('vendor/**') === -1) {
```
Fixed! I hope it makes it a bit clearer now.

mnapoli replied:
Perfect!

tomhatzer commented on index.js:
```diff
@@ -34,19 +39,210 @@ class ServerlessPlugin {
+        }
+
+        //let iamRoleStatements = this.serverless.service.provider.iamRoleStatements;
+        const roleDetails = {
```
Fixed!

tomhatzer commented on index.js:
```diff
@@ -34,19 +39,210 @@ class ServerlessPlugin {
+        let resources = this.serverless.service.provider.coreCloudFormationTemplate.Resources;
+        resources[resources.length] = {
+            'ServerlessDeploymentBucketLambdaAccessPolicy': roleDetails
+        };
```
It shouldn't replace it (I hope)... Basically it just appends an item to the cloud formation template resources, but I'm not 100% sure. I had no case where I could have tested this scenario. I tried doing this with the iamRoleStatements at first, but they were already parsed at that moment and I couldn't get the name of the deployment bucket and prefix without using references, but references are somehow not working after parsing the config.

If there's a way so I can use a reference outside of a `Fn::Join` I'd love to get some input/help here. 😅

mnapoli replied:
Yep I see what you mean but I don't see how this can work actually, the resource seems invalid because incorrectly formatted 🤔 

How about adding the permission to [iamRoleStatements](https://www.serverless.com/framework/docs/providers/aws/guide/iam/):

```js
this.serverless.service.provider.iamRoleStatements = this.serverless.service.provider.iamRoleStatements ? this.serverless.service.provider.iamRoleStatements : [];

this.serverless.service.provider.iamRoleStatements.push({
            Effect: 'Allow',
            Action: [
                's3:GetObject',
            ],
            Resource: [
                {
                    'Fn::Join': [
                        '',
                        [
                            'arn:aws:s3:::',
                            {
                                'Ref': 'ServerlessDeploymentBucket'
                            },
                            '/',
                            this.stripSlashes(this.provider.getDeploymentPrefix() + '/vendors/*')
                        ]
                    ]
                }
            ]
});
```

(haven't tested to be 100% transparent)

mnapoli replied:
Wait wait, now I'm wondering: at the `after:aws:deploy:deploy:createStack` step isn't the serverless config already parsed to produce the CloudFormation template?

Now I'm wondering if the vendor exclusion (and what I suggest in my comment above) are taken into account at all 🤔 

It may be worth moving those changes into the `package:setupProviderConfiguration` hook?

tomhatzer replied:
Yes, that sounds like a good plan, thank you. I'll add it right away. I had something like that in an earlier version but due to the problems with the `ServerlessDeploymentBucket` I removed it again. (That's why there was still a comment left for the `iamRoleStatements` selection.

I think this may be IMHO the best solution to add the rights for the s3 bucket. 👍

tomhatzer replied:
> Wait wait, now I'm wondering: at the `after:aws:deploy:deploy:createStack` step isn't the serverless config already parsed to produce the CloudFormation template?

Yes, it's already parsed, that's why I added it as a new element to the parsed template previously.

> Now I'm wondering if the vendor exclusion (and what I suggest in my comment above) are taken into account at all 🤔

In my tests this worked fine, I didn't have vendor exclusion in my serverless.yml file while running with separation enabled and the vendor folder was excluded fine.

> It may be worth moving those changes into the `package:setupProviderConfiguration` hook?

Ok I'll see if they are working inside that hook.

tomhatzer replied:
> It may be worth moving those changes into the package:setupProviderConfiguration hook?
@mnapoli we are already in that hook with this code. 😅

mnapoli replied:
Oh my bad!

If it works, then it works! 😅

tomhatzer commented on index.js:
```diff
@@ -34,19 +39,210 @@ class ServerlessPlugin {
+
+        this.consoleLog('Setting environment variables.');
+
+        this.serverless.service.provider.environment.BREF_DOWNLOAD_VENDOR = {
```
Fixed!

tomhatzer commented on index.js:
```diff
@@ -34,19 +39,210 @@ class ServerlessPlugin {
+            });
+        })
+            .then(() => {
+                const crypto = require('crypto');
```
Fixed!

tomhatzer commented on index.js:
```diff
@@ -34,19 +39,210 @@ class ServerlessPlugin {
+    }
+
+    async uploadVendorZip() {
+        await this.uploadZipToS3(this.filePath);
```
Good catch! I've added it here too. Thank you!

tomhatzer commented on index.js:
```diff
@@ -34,19 +39,210 @@ class ServerlessPlugin {
+        return path.replace(/^\/+/g, '');
+    }
+
+    async removeVendorArchives() {
```
Fixed!

tomhatzer commented on runtime/layers/console/breftoolbox.php:
```diff
@@ -0,0 +1,236 @@
+     * @param int $expires
+     * @return string
+     */
+    private static function AWS_S3_PresignDownload(
```
Fixed the code formatting and removed the annotations. I hope it's correct now.

tomhatzer commented:
Thank you very much for the review @mnapoli !

I've updated the code and added a few comments where necessary. 😃

tomhatzer commented on runtime/layers/console/breftoolbox.php:
```diff
@@ -0,0 +1,236 @@
+        $BucketName,
+        $AWSRegion,
+        $canonical_uri,
+        $expires = 86400
```
Fixed!

mnapoli reviewed (approved):
Awesome! Looking good!

Have you tested that latest version? I should take some time ASAP to give it a quick test and merge + release (I'd rather avoid merging until I'm ready to release for large changes like these).

tomhatzer commented:
Thank you very much for the review @mnapoli ! I'm going to test this in the next few hours and see if it still works like intended. Will get back to you with test results 👍

tomhatzer commented:
@mnapoli I've made a small change for the iamRole expansion. I've moved it to the `before:package:setupProviderConfiguration` hook including the exclusion of the `vendor/**` path. Now the rights are included in the stack creation itself and not added afterwards.

I've also added a check if the bucket still exists when removing the stack.

Testing was successful, stack was created fine and removed without traces too.

mnapoli commented:
Hey Tom, very sorry for the delay, with the holidays it's been a while.

Let's merge this! Thanks a lot for the time you spent on this!
